buildengine/engine.go

```diff
@@ -496,13 +498,115 @@ func (e *Engine) buildWithCallback(ctx context.Context, callback buildCallback,
 		allErrors = append(allErrors, err)
 	}
 
+	allErrors = append(allErrors, e.validateConfigsAndSecretsMatch(ctx, builtModules)...)
+
 	if len(allErrors) > 0 {
 		return errors.Join(allErrors...)
 	}
 
 	return nil
 }
 
+func (e *Engine) validateConfigsAndSecretsMatch(ctx context.Context, builtModules map[string]*schema.Module) []error {
+	errs := []error{}
+	logger := log.FromContext(ctx)
+
+	configsProvidedGlobally := make(map[string]bool)
```

Comment:
There's quite a lot of duplication here. I think this could be refactored to include a reusable function that takes a map of merged global+module entries and compares them to a map of used module entries, and returns errors with the differences. Pass in the "type" of entries (config or secret) as a string for more useful errors.

Reply:
Keys in the merged map could be in the form `[<module>.]<key>`